### ISSUE TITLE
Multiple Flutter Engine: destroy flutter engine making live flutter stop response user interaction

### ISSUE THREAD

**Issue by zhangxu-mag** [flutter/flutter#76597]
## Steps to Reproduce

1. User FlutterEngineGroup create new flutter engine
2. Say we have Activity1 associated with flutter engine 1, Activity2 associated with flutter engine 2
3. when user open the app, Activity1 is created and it create  flutter engine 1 to render the view
4. user interact with Activity1 to open Activity2
5. Activity2 is created and it created flutter engine 2 to render view
6. user navigate back
7. Activity2 is destroy triggering flutter enige 2 destroy
8. User back to Activity1
9. User interact with Activity1 and no any response

**Expected results:** 
User can continue interact with Activity1

**Actual results:** 
User can not continue interact with Activity1

<details>
comments out "flutterJNI.detachFromNativeAndReleaseResources();" in FlutterEngine.destory helps fix      @@@@@this.

</details>

**Comment by zhangxu-mag** [comment]
@gaaclarke May you help take a look.

@xster FYI.

**Comment by gaaclarke** [comment]
@zhangxu-mag This is a flow that works for me, see the sample code: https://github.com/flutter/samples/pull/677 ([commit](https://github.com/gaaclarke/samples/commit/cf6b06332101255394d57654adbe6c4cb422dbc2))

Please take a look to see what you are doing different than that sample.  If you could provide a patch to that sample to reproduce your problem that would be helpful.

**Comment by zhangxu-mag** [comment]
From the example code, I did not see you have explicit destroy the engine. (Correct if I am wrong, not family with kt).

Sorry I try on my own example, so hard to provide a patch.

**Comment by TahaTesser** [comment]
Hi @zhangxu-mag 
Can you please provide your `flutter doctor -v` and `flutter run --verbose` logs, and a [minimal complete reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example)?
Thank you

**Comment by gaaclarke** [comment]
@zhangxu-mag I'm more familiar with iOS development, under what conditions would you explicitly call destroy on a FlutterEneign versus letting the garbage collector free it?

**Comment by xster** [comment]
Destroy should always be called explicitly and the garbage collector finalizers should generally not be relied on in Java. We should add a destroy call in the sample as well. Maybe in https://github.com/flutter/samples/pull/677/files#diff-2afc2a0aeb550ee843dbe638b73927aa4e45da4b0ece8549d3779d343cc58243R70.

**Comment by gaaclarke** [comment]
Thanks xster.  I introduced the `FlutterEngine.destroy()` and I'm seeing a crash as a result.  I'm looking into it but running into problems running the sample with a local engine.

Steps to reproduce:
https://github.com/flutter/samples/pull/677

```diff
index ad14780..e40e959 100644
--- a/add_to_app/multiple_flutters/multiple_flutters_android/app/src/main/java/dev/flutter/multipleflutters/EngineBindings.kt
+++ b/add_to_app/multiple_flutters/multiple_flutters_android/app/src/main/java/dev/flutter/multipleflutters/EngineBindings.kt
@@ -74,6 +74,7 @@ class EngineBindings(activity: Activity, delegate: EngineBindingsDelegate, entry
      * This tears down the messaging connections on the platform channel and the DataModel.
      */
     fun detach() {
+        engine.destroy()
         DataModel.instance.removeObserver(this)
         channel.setMethodCallHandler(null)
     }
```

**Comment by gaaclarke** [comment]
If I run the project from android studio it stops responding to touch after popping the FlutterActivity, but if I run the app through `flutter run` it crashes with the following crash:
```
02-25 13:18:36.268 14292 14292 F flutter : [FATAL:flutter/fml/memory/weak_ptr.h(262)] Check failed: (checker_.checker).IsCreationThreadCurrent(). 
02-25 13:18:36.268 14292 14292 F libc    : Fatal signal 6 (SIGABRT), code -6 in tid 14292 (example.regular)
02-25 13:18:36.275   469   469 W         : debuggerd: handling request: pid=14292 uid=10221 gid=10221 tid=14292
02-25 13:18:36.345 14385 14385 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
02-25 13:18:36.345 14385 14385 F DEBUG   : Build fingerprint: 'asus/WW_ASUS_A002/ASUS_A002:7.0/NRD90M/14.1200.1703.110-20170325:userdebug/release-keys'
02-25 13:18:36.345 14385 14385 F DEBUG   : Revision: '0'
02-25 13:18:36.345 14385 14385 F DEBUG   : ABI: 'arm'
02-25 13:18:36.345 14385 14385 F DEBUG   : pid: 14292, tid: 14292, name: example.regular  >>> com.example.regular <<<
02-25 13:18:36.345 14385 14385 F DEBUG   : signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
02-25 13:18:36.349 14385 14385 F DEBUG   : Abort message: '[FATAL:flutter/fml/memory/weak_ptr.h(262)] Check failed: (checker_.checker).IsCreationThreadCurrent(). 
02-25 13:18:36.349 14385 14385 F DEBUG   : '
02-25 13:18:36.349 14385 14385 F DEBUG   :     r0 00000000  r1 000037d4  r2 00000006  r3 00000008
02-25 13:18:36.350 14385 14385 F DEBUG   :     r4 f366a58c  r5 00000006  r6 f366a534  r7 0000010c
02-25 13:18:36.350 14385 14385 F DEBUG   :     r8 ffa41040  r9 f0a05400  sl 00000000  fp ffa40fcc
02-25 13:18:36.350 14385 14385 F DEBUG   :     ip 00000009  sp ffa40ad0  lr f1344527  pc f1346da8  cpsr 200f0010
02-25 13:18:36.352 14385 14385 F DEBUG   : 
02-25 13:18:36.352 14385 14385 F DEBUG   : backtrace:
02-25 13:18:36.358 14385 14385 F DEBUG   :     #00 pc 00049da8  /system/lib/libc.so (tgkill+12)
02-25 13:18:36.358 14385 14385 F DEBUG   :     #01 pc 00047523  /system/lib/libc.so (pthread_kill+34)
02-25 13:18:36.359 14385 14385 F DEBUG   :     #02 pc 0001d779  /system/lib/libc.so (raise+10)
02-25 13:18:36.359 14385 14385 F DEBUG   :     #03 pc 00019271  /system/lib/libc.so (__libc_android_abort+34)
02-25 13:18:36.359 14385 14385 F DEBUG   :     #04 pc 000172d8  /system/lib/libc.so (abort+4)
02-25 13:18:36.359 14385 14385 F DEBUG   :     #05 pc 0000453f  /data/app/com.example.regular-2/lib/arm/libflutter.so (offset 0x1587000)
```

**Comment by gaaclarke** [comment]
I'm having a hard time getting a debugger attached and symbolicated but I've been able to add some print statements to pinpoint the problem:
```diff
diff --git a/shell/common/shell.cc b/shell/common/shell.cc
index 07c4172c6..d90f23ab7 100644
--- a/shell/common/shell.cc
+++ b/shell/common/shell.cc
@@ -437,14 +438,22 @@ Shell::~Shell() {
 
   fml::AutoResetWaitableEvent ui_latch, gpu_latch, platform_latch, io_latch;
 
-  fml::TaskRunner::RunNowOrPostTask(
-      task_runners_.GetUITaskRunner(),
+  FML_LOG(ERROR) << "foobaz 2";
+  const auto& ui_task_runner = task_runners_.GetUITaskRunner();
+  FML_LOG(ERROR) << "foobaz 2.1";
+  auto task =
       fml::MakeCopyable([engine = std::move(engine_), &ui_latch]() mutable {
+        FML_LOG(ERROR) << "foobaz 2.4";
         engine.reset();
+        FML_LOG(ERROR) << "foobaz 2.5";
         ui_latch.Signal();
-      }));
+      });
+  FML_LOG(ERROR) << "foobaz 2.2";
+  fml::TaskRunner::RunNowOrPostTask(ui_task_runner, task);
+  FML_LOG(ERROR) << "foobaz 2.3";
   ui_latch.Wait();
```

When popping the activity with 2 flutter instances the trace is:
```
E/flutter (28131): [ERROR:flutter/shell/common/shell.cc(443)] foobaz 2.1
E/flutter (28131): [ERROR:flutter/shell/common/shell.cc(451)] foobaz 2.2
E/flutter (28131): [ERROR:flutter/shell/common/shell.cc(453)] foobaz 2.3
E/flutter (28131): [ERROR:flutter/shell/common/shell.cc(446)] foobaz 2.4
E/flutter (28131): [ERROR:flutter/shell/common/shell.cc(448)] foobaz 2.5
E/flutter (28131): [ERROR:flutter/shell/common/shell.cc(443)] foobaz 2.1
E/flutter (28131): [ERROR:flutter/shell/common/shell.cc(451)] foobaz 2.2
E/flutter (28131): [ERROR:flutter/shell/common/shell.cc(453)] foobaz 2.3
<hang or crash>
```

So, it looks like the first shell is deallocating correctly.  The second shell is getting hung in the first latch for the UI task runner.  I suspect the first dealloc is trashing the threadhost instead of just releasing its reference to the shared threadhost.

**Comment by gaaclarke** [comment]
I found the fix for the hang.  It was what I suspected, clearing an object instead of a shared reference:
```diff
diff --git a/shell/platform/android/android_shell_holder.cc b/shell/platform/android/android_shell_holder.cc
index 6a71c24e2..6e17072a6 100644
--- a/shell/platform/android/android_shell_holder.cc
+++ b/shell/platform/android/android_shell_holder.cc
@@ -170,7 +170,7 @@ AndroidShellHolder::AndroidShellHolder(
 
 AndroidShellHolder::~AndroidShellHolder() {
   shell_.reset();
-  thread_host_->Reset();
+  thread_host_.reset();
 }
 ```

I get a crash now because the FlutterRenderer and the FlutterEngine share an instance of FlutterJNI, since the engine already cleaned it up we get the following exception when the FlutterRenderer is cleaned up:
```
E/AndroidRuntime(29145): Caused by: java.lang.RuntimeException: Cannot execute operation because FlutterJNI is not attached to native.
E/AndroidRuntime(29145): 	at io.flutter.embedding.engine.FlutterJNI.ensureAttachedToNative(FlutterJNI.java:386)
E/AndroidRuntime(29145): 	at io.flutter.embedding.engine.FlutterJNI.onSurfaceDestroyed(FlutterJNI.java:497)
E/AndroidRuntime(29145): 	at io.flutter.embedding.engine.renderer.FlutterRenderer.stopRenderingToSurface(FlutterRenderer.java:224)
E/AndroidRuntime(29145): 	at io.flutter.embedding.android.FlutterView.detachFromFlutterEngine(FlutterView.java:993)
E/AndroidRuntime(29145): 	at io.flutter.embedding.android.FlutterActivityAndFragmentDelegate.onDestroyView(FlutterActivityAndFragmentDelegate.java:492)
E/AndroidRuntime(29145): 	at io.flutter.embedding.android.FlutterFragment.onDestroyView(FlutterFragment.java:667)
E/AndroidRuntime(29145): 	at androidx.fragment.app.Fragment.performDestroyView(Fragment.java:2811)
E/AndroidRuntime(29145): 	at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManagerImpl.java:954)
E/AndroidRuntime(29145): 	at androidx.fragment.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManagerImpl.java:1238)
E/AndroidRuntime(29145): 	at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManagerImpl.java:1303)
E/AndroidRuntime(29145): 	at androidx.fragment.app.FragmentManagerImpl.dispatchStateChange(FragmentManagerImpl.java:2659)
E/AndroidRuntime(29145): 	at androidx.fragment.app.FragmentManagerImpl.dispatchDestroy(FragmentManagerImpl.java:2644)
E/AndroidRuntime(29145): 	at androidx.fragment.app.FragmentController.dispatchDestroy(FragmentController.java:329)
E/AndroidRuntime(29145): 	at androidx.fragment.app.FragmentActivity.onDestroy(FragmentActivity.java:366)
E/AndroidRuntime(29145): 	at com.example.regular.DoubleFlutterActivity.onDestroy(DoubleFlutterActivity.kt:74)
```

It almost seems like no one was ever destroying the engine explicitly.  I'm continuing to investigate.

**Comment by gaaclarke** [comment]
I got rid of the crash by making sure to call destroy after the surface has been destroyed.   I'll land the fix and the change to the sample code:
```diff
diff --git a/add_to_app/multiple_flutters/regular/android/app/src/main/kotlin/com/example/regular/DoubleFlutterActivity.kt b/add_to_app/multiple_flutters/regular/android/app/src/main/kotlin/com/example/regular/DoubleFlutterActivity.kt
index bb636ec..4061bd3 100644
--- a/add_to_app/multiple_flutters/regular/android/app/src/main/kotlin/com/example/regular/DoubleFlutterActivity.kt
+++ b/add_to_app/multiple_flutters/regular/android/app/src/main/kotlin/com/example/regular/DoubleFlutterActivity.kt
@@ -64,14 +64,14 @@ class DoubleFlutterActivity : FragmentActivity(), EngineBindingsDelegate {
     }
 
     override fun onDestroy() {
-        topBindings.detach()
-        topBindings.detach()
-
         for (i in 0 until numberOfFlutters) {
             FlutterEngineCache.getInstance().remove(i.toString())
         }
 
         super.onDestroy()
+
+        topBindings.detach()
+        bottomBindings.detach()
     }
```

**Comment by xster** [comment]
Thanks for the fix. We might want to consider bundling into a hotfix.

**Comment by gaaclarke** [comment]
> Thanks for the fix. We might want to consider bundling into a hotfix.

Done, submitted for cp 1.26

**Comment by TahaTesser** [comment]
@xster 
Thanks for the labels, so sorry I missed that.
